Documentation: how to use auto monitor feature to keep the chip alive (#2543)
apache/incubator-nuttx: Documentation/components/drivers/character/watchdog.rst

.. c:macro:: WDIOC_KEEPALIVE

>  This command resets the watchdog timer ("ping",  "pet the dog",  "feed the dog"). 

Reviewer: ```suggestion
 This command resets the watchdog timer AKA '**ping**", "**kick**", "**pet** the dog",  "**feed** the dog"). 
```

>  This command resets the watchdog timer AKA '**ping**", "**kick**", "**pet** the dog",  "**feed** the dog"). 

Reviewer: ```suggestion
 This command resets the watchdog timer AKA '**ping**", "**kick**", "**pet**",  "**feed**" the dog". 
```

> Using the Auto-monitor feature to keep the chip alive

Reviewer: I am confused by this. Is this named correctly? A HW watchdog needs to be "reset"  if not when it times out it is "triggered".  

So what is  Auto-monitor? Is the generating the keep-alive? Then it is not a monitor,

Author: Hey @davids5
What name do you refer to?
The `Auto-monitor` is an option that is under the Watchdog Timer Support option.
The `Auto-monitor` provides 4 mechanisms to feed the dog automatically.
I think the `auto` prefix stands for automatically,i.e., without the developer configuring it in execution code and the `monitor` word means that it will monitor your code because if one task gets stuck the timer will expire and the chip will be reseted.

Reviewer: Ah, I get it:  It is really  built in monitoring the system and resetting the watchdog.  "Enable Built in System Monitoring to reset the watchdog."

Is the timer one a HW timer off an ISR?  Because WD kicking done off an ISR is a really bad idea.

Author: "Enable Built in System Monitoring to reset the watchdog." is a better title indeed.

The timer option calls `wd_start` in the beginning. 

> This function adds a watchdog timer to the active timer queue.

 and registers a function that will be called from the interrupt level after the specified number of ticks has elapsed. 
When this option is selected, the registered function resets the watchdog and calls `wd_start` again registering itself again.

It is implemented in `drivers/timers/watchdog.c`

Reviewer: > "Enable Built in System Monitoring to reset the watchdog." is a better title indeed.
I am easily confused but what I read a lot so I am glad it makes sense, 

> registers a function that will be called from the interrupt level after the specified number of ticks has elapsed.

Yikes, I would put a caution all over that to not to not use it, less one really knows what they are doing.,,,,,

> Using the Auto-monitor feature to keep the chip alive

Reviewer: ```suggestion
Using the Auto-monitor feature to reset the watch dog,
```

------------------------------------------------------

> The auto-monitor provides an OS-internal mechanism to automatically start the watchdog timer and repeatedly reset the counting after the watchdog is registered.

Reviewer: ```suggestion
The auto-monitor provides an OS-internal mechanism to automatically start and repeatedly reset the watchdog.  
```

Reviewer: I would replace the fed, ping, pet, with the word reset as defined above

2. Enable the Auto-monitor option

 Go into menu :menuselection:`Device Drivers --> Timer Driver Support` and press :kbd:`Enter`. Then enable:

 - [x] Watchdog Timer Support

 Then press :kbd:`Enter` again to enter into the Watchdog Timer Support menu. And finally enable the Auto-monitor option:

 - [x] Auto-monitor

 After selecting the option you may want to configure some parameters:

 * **Timeout**: It is the watchdog timer expiration time in seconds.
 * **Keep a live interval**: This is the interval in which the watchdog will be fed. It is in seconds. It can't be bigger than the timeout. If this interval is equal to timeout interval, than this interval will automatically change to half timeout. 
 * **Keep alive by**: This is a choice to determine who is going to feed the dog. There are 4 possible choices that are described as follows.

 ``Capture callback``: This choice registers a watchdog timer callback to reset the watchdog every time it expires, i.e., on timeout. 

 ``Timer callback``: This choice also uses a timer callback to reset the watchdog, but it will reset the watchdog every "keep a live interval".

>  ``Worker callback``: This choice uses the Low Priority Work Queue to reset the watchog every "keep a live interval". This choice depends on having the Low Priority Work Queue enabled. 

Reviewer 2: not really need Low Priority Work Queue:

1. If only high work enable, the high work queue is used
2. otherwise the low work queue is used

This is because, high work queue will map to low work queue if only high work queue is enabled.

Author: Hum, so I'll change and add it as follows:

``Worker callback``: This choice uses a Work Queue to reset the watchdog every "keep a live interval". This choice depends on having the Low or High Priority Work Queue enabled.
If only the High Priority Work Queue is enabled, this one will be used, otherwise Low Priority Work Queue is used.

What do you think?

BTW, I thought it was only LPWORK because of the configuration:
```
      work_queue(LPWORK, &upper->work, watchdog_automonitor_worker,
                 upper, WATCHDOG_AUTOMONITOR_PING_INTERVAL);
```

I didn't know the LPWORK was replaced by the HPWORK in case the latter was enabled and the former was not.

Reviewer: Why invent terms? Keep alive -> Watchdog Reset interval

Reviewer 2: @saramonteiro yes, it's accurate now, thanks.

Author: @davids5 I did not create terms like "Keep alive" and "pet the dog" or "ping". 
Actually I just used the terms I saw in `drivers/timers/watchdog.c` and in `include/nuttx/timers/watchdog.h`
For example, the ioctl that feeds the dog is actually named `WDIOC_KEEPALIVE`.
My main goal with this PR is to document the WDT related features sticking to the current terms used in the source code.

Author: @xiaoxiang781216 thank you! =)

Reviewer: > @davids5 I did not create terms like "Keep alive" and "pet the dog" or "ping".
> Actually I just used the terms I saw in `drivers/timers/watchdog.c` and in `include/nuttx/timers/watchdog.h`
> For example, the ioctl that feeds the dog is actually named `WDIOC_KEEPALIVE`.
> My main goal with this PR is to document the WDT related features sticking to the current terms used in the source code.

Fare enough.
